contracts/contracts/governance/L2Governance.sol

```diff
@@ -0,0 +1,498 @@
+        );
+
+        return proposalId;
+    }
+
+    /**
+     * @dev Creates a proposal with the given args.
+     *      Can be called by anyone
+     */
+    function propose(
+        address[] memory targets,
+        uint256[] memory values,
+        string[] memory signatures,
+        bytes[] memory calldatas,
+        string memory description
+    ) external virtual returns (uint256) {
+        return
+            _storeProposal(
+                msg.sender,
+                targets,
+                values,
+                signatures,
+                calldatas,
+                description
+            );
+    }
+
+    /**
+     * @dev Queues a proposal on the Timelock
+     *      Private and only to be used by Mainnet Governance through CCIP Router
+     */
+    function _queue(uint256 proposalId) internal {
+        TimelockController controller = TimelockController(payable(timelock));
+        ProposalDetails memory details = proposalDetails[proposalId];
+
+        if (!details.exists) {
+            revert InvalidProposal();
+        }
+
+        controller.scheduleBatch(
+            details.targets,
+            details.values,
+            _encodeCalldata(details.signatures, details.calldatas),
+            0,
+            details.descriptionHash,
+            controller.getMinDelay()
+        );
+
+        emit ProposalQueued(proposalId);
+    }
+
+    /**
+     * @dev Cancels a pending proposal on the Timelock
+     *      Private and only to be used by Mainnet Governance through CCIP Router
+     */
+    function _cancel(uint256 proposalId) internal {
+        TimelockController controller = TimelockController(payable(timelock));
+
+        bytes32 timelockHash = _getTimelockHash(proposalId);
+
+        proposalDetails[proposalId].exists = false;
+        controller.cancel(timelockHash);
+
+        emit ProposalCanceled(proposalId);
+    }
+
+    /**
+     * @dev Returns the timelock hash for a proposal
+     */
+    function _getTimelockHash(uint256 proposalId)
+        internal
+        view
+        returns (bytes32 timelockHash)
+    {
+        TimelockController controller = TimelockController(payable(timelock));
+
+        ProposalDetails memory details = proposalDetails[proposalId];
+
+        if (!details.exists) {
+            revert InvalidProposal();
+        }
+
+        timelockHash = controller.hashOperationBatch(
+            details.targets,
+            details.values,
+            _encodeCalldata(details.signatures, details.calldatas),
+            0,
+            details.descriptionHash
+        );
+    }
+
+    function getTimelockHash(uint256 proposalId)
+        external
+        view
+        returns (bytes32)
+    {
+        return _getTimelockHash(proposalId);
+    }
+
+    /**
+     * @dev Executes an already queued proposal on the Timelock.
+     *      Can be called by anyone
+     * @param proposalId Proposal ID
+     */
+    function execute(uint256 proposalId) external {
```

Comment:
It might be advisable to verify the status of the Chainlink bridge as outlined [here(https://docs.chain.link/ccip/concepts#offchain-risk-management-node). For instance, on the Optimism network, the following contract is typically referred to for this purpose: [View Contract on Optimistic Etherscan](https://optimistic.etherscan.io/address/0x8C7C2C3362a42308BB5c368677Ad321D11693b81#readContract). However, this step may not be essential unless there is an issue with the bridge when a CANCEL_PROPOSAL_COMMAND is sent just before the potential execution.

Reply:
Will do. However, even if a `CANCEL_PROPOSAL_COMMAND` is sent by a compromised bridge, it'll not be immediately executable, since it has to go through the timelock

Reply:
Ah, you're right. My bad, I got confused